docs/rewriter/design.md

```diff
@@ -0,0 +1,130 @@
+ONNX Script complains about a missing required-attribute, that would
+be a problem when specifying a pattern.
+
+Option 2 allows us to have a more extensible mechanism, customized to
+the case of patterns. For example, it is easy to allow users to specify
+that we are looking for attribute `alpha` to have a value of `0.16325`
+within some error tolerance limits. (Note that we can push such conditions
+into the guard/condition function too. So, this specific example is more
+of a syntactic-sugar convenience.) Another interesting use-case, which
+comes up in practice, is when we want to do some wild-card matching against
+a op/function name ... eg., if we want to match against a call to a
+function (produced by the pytorch exporter) with a name that has `Silu` as
+a prefix.
+
+## IR and Builder for Patterns and Rewrite-Rules
+
+We already have a couple of approaches to pattern-matching based rewrite-rules:
+the original one (restricted to single-output patterns, but supporting other
+features such as a mechanism to bind variables to specific nodes of the
+pattern, and using them to check for extra-conditions before applying a
+rewrite-rule, and allowing for more efficient match-and-replace), and a
+newer extension to support multiple-output patterns, but not integrated
+with the other mechanisms.
+
+We can integrate and simplify these approaches, including the different options
+for describing patterns mentioned above, by unifying the classes defining
+the patterns and rewrite-rules (into a pattern IR) and by providing different
+factory (constructor) methods for patterns and rewrite-rules. Specifically,
+these methods will be the developer-facing APIs for expressing rewrite-rules.
+Hence, it is important for us to consolidate these APIs to avoid potential
+confusion among users writing rewrite-rules. (The implementations of the
+pattern-matching can vary, depending on the complexity of the pattern,
+such as whether it is single-output or multiple-output, etc.)
+
+Proposal:
+
+* The rewrite-rule constructor takes two parameters, one to specify the
+pattern, and one to specify the replacement. We can provide a version with
+an extra condition parameter (as a convenience feature).
+
+* The pattern can be either a FunctionProto (or, equivalently, an ONNX Script
+function) or a function that constructs a pattern-graph or a pattern-graph.
+In the first two cases, they are converted into a pattern-graph.
+
+* The replacement is a function that accepts an expanded match-bindings
+(that is, `**match_bindings`) and returns either `None` (in case of failure)
+or a representation of the modification to the graph (in terms of removed
+and added nodes/values).
+
+* We may benefit from using a trace-mode onnxscript function above. In
+particular, we may be able to improve upon the replacement function by
+integrating the graph-modification cleanly into a generalization of the
+trace-mode onnxscript.
+
+
```

Comment:
I would add an item about the information the user needs to understand why a pattern did not work as expected. It is better to design it at the beginning than add it later.

Reply:
That would be great. Any concrete suggestions? Would it be just information printed/logged (controlled by parameters like `verbose` in your implementation, or whenever a match fails in the middle, etc.)?

Reply:
I implemented a mechanism which helped me to understand when a match fails which nodes of pattern are matched, why the match fail and where. The example i added illustrates what the user sees when a match fails.